apps/studio/src/server/modules/rate-limit/rate-limit.service.ts

```diff
@@ -0,0 +1,52 @@
+import { type NextApiRequest } from "next"
+import { PrismaClient } from "@prisma/client"
+import { TRPCError } from "@trpc/server"
+import {
+  RateLimiterMemory,
+  RateLimiterPrisma,
+  RateLimiterRes,
+} from "rate-limiter-flexible"
+
+import { type RateLimitMetaOptions } from "./types"
+import { getRateLimitFingerprint } from "./utils"
+
+// Default 5 queries per second fallback
+export const rateLimiterMemory = new RateLimiterMemory({
+  points: 5,
+  duration: 1,
+})
+
+export async function checkRateLimit({
+  rateLimitOptions,
+  req,
+  prisma,
+}: {
+  rateLimitOptions: RateLimitMetaOptions
+  req: NextApiRequest
+  prisma: PrismaClient
+}) {
+  const max = rateLimitOptions.max ?? 5
+  const windowMs = rateLimitOptions.windowMs ?? 1000
+
+  const store = new RateLimiterPrisma({
```

Comment:
i think we can store this in memory; don't need to put in the DB. See morpho codebase!

Reply:
i took this with reference to morpho's codebase - in there, they used a redis instance, which is also shared memory between their compute, with the insurance limiter being in memory.

i think a shared storage is needed because we want to avoid the case where different requests go to different instances and lead to different attempts being logged in the memory of each instance.

for example, a user attempts to verify their otp and it goes to instance A. it fails and they retry, causing the second attempt to go to instance B. in this case, if we use a memory backed rate limiter, the isntances would each have count: 1 rather than the actual count of 2

Reply:
yes needs a shared instance - hence a cache or DB is required